New background correction option and improved junge calculation

arsalanmstn opened the pull request:
The 'get_j' function is now modified to handle zero_value RuntimeWarning encountered in np.log(number_distribution[ind])

nepstad commented:
This is a good addition. The new background correction method should also be possible to set from the config file. We should consider adding the required logic in this pull request, and updating the documentation.

nepstad commented on pyopia/background.py:
```diff
@@ -151,6 +151,44 @@ def correct_im_fast(imbg, imraw):
+        for i in range(imbg.shape[2]):
+            imbg[:, :, i][imbg[:, :, i] == 0] = 1 / 255
+        else:
+            imbg[imbg == 0] = 1 / 255
```
I think you can replace all the preceding lines with this: 

imbg = np.clip(imbg, a_min=1/255)

nepstad commented on pyopia/background.py:
```diff
@@ -284,7 +327,10 @@ def __call__(self, data):
+        if self.divide_bg:
+            data['im_corrected'] = correct_im_accurate_divide(data['imbg'], data[self.image_source])
+        else:
+            data['im_corrected'] = correct_im_accurate(data['imbg'], data[self.image_source])
```
Perhaps it would be simpler to pass a flag to the existing "correct_im_accurate" to indicate which background correction method to use? Then have the if inside that function. This avoids the need for a new function with some duplicate code.

arsalanmstn replied:
Nice, I will revise the code

nepstad reviewed (changes requested):
Looks good! I suggest also updating the background notebook in the docs/ folder with a simple example of how to use this new functionality, and perhaps compare the result of division and subtraction approach.

emlynjdavies reviewed (changes requested):
suggest this is version 2.6.0, since it introduces a new functionality

arsalanmstn commented:
> suggest this is version 2.6.0, since it introduces a new functionality

Hei @emlynjdavies,

Would you please add couple of empty images (images with zero particles) into oil.zip [pyopia.exampledata.get_file_from_pysilcam_blob('oil.zip')] for background correction? In this case we can show the differences between subtraction and division in a nicer way in the background notebook in the docs/ folder. The current notebook contains white-color circles in the corrected image!!!

emlynjdavies commented:
> > suggest this is version 2.6.0, since it introduces a new functionality
> 
> Hei @emlynjdavies,
> 
> Would you please add couple of empty images (images with zero particles) into oil.zip [pyopia.exampledata.get_file_from_pysilcam_blob('oil.zip')] for background correction? In this case we can show the differences between subtraction and division in a nicer way in the background notebook in the docs/ folder. The current notebook contains white-color circles in the corrected image!!!

The oil directory is intended to just contain oil images. You can use a few (e.g. 10) of the oil images to create the background (doesn't need to be clean images, just as long as you have enough moving droplets that they average out ok)

arsalanmstn commented:
Hei all,

I made some changes in correct_im_accurate()/background.py. I realized that if the background is not clean enough, the division mode would produce a corrected image out of the (0,1) range-a dark image!!!. So, I used couple of extra lines to handle the problem (clipping and shifting negative values to zero for division mode). The background notebook is now modified to show the two different methods of subtraction and division, however, for some reasons, it does not pass the pipeline_holo_test(). I had the same problem when I passed the divide_bg = True as the default value, and changed it to False. I also had to change the version in __init__.py from '2.6.2' to '2.6.3' to pass the version check.

emlynjdavies commented:
> Hei all,
> 
> I made some changes in correct_im_accurate()/background.py. I realized that if the background is not clean enough, the division mode would produce a corrected image out of the (0,1) range-a dark image!!!. So, I used couple of extra lines to handle the problem (clipping and shifting negative values to zero for division mode). The background notebook is now modified to show the two different methods of subtraction and division, however, for some reasons, it does not pass the pipeline_holo_test(). I had the same problem when I passed the divide_bg = True as the default value, and changed it to False. I also had to change the version in **init**.py from '2.6.2' to '2.6.3' to pass the version check.

Excellent. I think it is best that @nepstad take a final review on this one ;)

nepstad commented on pyopia/background.py:
```diff
@@ -108,17 +110,26 @@ def correct_im_accurate(imbg, imraw):
+        imbg = np.clip(imbg, a_min=1/255, a_max=None)   # Clipping the zero_value pixels
+        im_corrected = imraw / imbg
+        im_corrected += (1 / 2 - np.percentile(im_corrected, 50))
+        im_corrected -= im_corrected.min()   # Shift the negative values to zero
```
This line is doing the same thing as the next one I think, and could be removed.

nepstad replied:
The comment is somewhat misleading - this operation shifts the whole intensity range by the minimum value.

nepstad commented:
> suggest this is version 2.6.0, since it introduces a new functionality

The version should be changed to 2.7.0, since the main branch is on 2.6.2 now.

nepstad commented on pyopia/__init__.py:
```diff
@@ -1 +1 @@
-__version__ = '2.6.2'
+__version__ = '2.6.3'
```
Change version to 2.7.0
